Let logical_xor support non-bool tensors.

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24379 Implement operator logical_and
* **#23978 Let logical_xor support non-bool tensors.**
* #23916 Let logical_not support non-bool tensors.
* #23860 Recommend logical_not() instead of bitwise_not() when applying sub and neg on bool tensors.
* #23847 Add logical_xor operator
* #23839 Add logical_not operator.

Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)

xuhdev committed:
Let logical_xor support non-bool tensors.

xuhdev committed:
Update on "Let logical_xor support non-bool tensors."


Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)

gchanan commented:
@xuhdev: can you rebase this entire stack?  There's a merge conflict with master.

xuhdev committed:
Update on "Let logical_xor support non-bool tensors."


Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)

xuhdev commented:
@gchanan Done!

xuhdev commented:
I'll work on `logical_and` and `logical_or` to make logical operation complete once this stack is approved (so that we have a consensus on the way to deal with non-bool operands).

gchanan reviewed (changes requested):
same comments as logical_not, I think we should support any output dtype.

xuhdev committed:
Update on "Let logical_xor support non-bool tensors."


Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)

xuhdev committed:
Update on "Let logical_xor support non-bool tensors."


Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)

xuhdev committed:
Update on "Let logical_xor support non-bool tensors."


Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)

xuhdev committed:
Update on "Let logical_xor support non-bool tensors."


Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)

xuhdev committed:
Update on "Let logical_xor support non-bool tensors."


Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)

xuhdev committed:
Update on "Let logical_xor support non-bool tensors."


Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)

xuhdev committed:
Update on "Let logical_xor support non-bool tensors."


Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)

xuhdev commented:
@gchanan Updated!

gchanan commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -81,11 +81,29 @@ void div_kernel(TensorIterator& iter) {
-  cpu_kernel(iter,
-    [](bool a, bool b) -> bool {
-      return a != b;
+  AT_DISPATCH_ALL_TYPES_AND(kBool, iter.dtype(1), "logical_xor_cpu", [&]() {
```
I think we should support half here.

gchanan commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -81,11 +81,29 @@ void div_kernel(TensorIterator& iter) {
 }
 
+template <class self_t, class other_t>
+void logical_xor_kernel_impl(TensorIterator& iter) {
```
is this used?

gchanan commented on torch/_torch_docs.py:
```diff
@@ -2570,18 +2570,33 @@ def merge_dicts(*dicts):
     >>> torch.logical_xor(torch.tensor([True, False, True]), torch.tensor([True, False, False]))
     tensor([ False, False,  True])
-""")
+    >>> a = torch.tensor([0, 1, 10, 3], dtype=torch.int8)
```
nit: the latest column should probably be 0,0 to show all the cases.

gchanan commented on torch/_torch_docs.py:
```diff
@@ -2570,18 +2570,33 @@ def merge_dicts(*dicts):
+    tensor([ True,  True, False, False])
+    >>> torch.logical_xor(a.double(), b)
+    tensor([ True,  True, False, False])
+    >>> torch.logical_xor(a, b, out=torch.empty(0, dtype=torch.bool))
```
same as the logical not review, we should pass in the correct size.

gchanan commented on torch/_torch_docs.py:
```diff
@@ -2570,18 +2570,33 @@ def merge_dicts(*dicts):
+    tensor([ True,  True, False, False])
+    >>> torch.logical_xor(a, b, out=torch.empty(0, dtype=torch.int8))
+    tensor([1, 1, 0, 0], dtype=torch.int8)
+    >>> torch.logical_xor(a, b, out=torch.empty(0, dtype=torch.double))
```
as with the logical_not, we don't need all these cases, just one non-bool non-out, one non-bool out variant is sufficient.

gchanan commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -81,11 +81,29 @@ void div_kernel(TensorIterator& iter) {
+  AT_DISPATCH_ALL_TYPES_AND(kBool, iter.dtype(0), "logical_xor_cpu", [&]() {
+    cpu_kernel(iter,
+      [](self_t a, other_t b) -> scalar_t {
+        return static_cast<scalar_t>((!a) != (!b));
```
this seems fine, just curious why you didn't convert to bool rather than using "!"?

xuhdev replied:
I guess I got some foreign language influence :) Fixed now.

xuhdev committed:
Update on "Let logical_xor support non-bool tensors."


Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)

xuhdev committed:
Update on "Let logical_xor support non-bool tensors."


Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)

xuhdev committed:
Update on "Let logical_xor support non-bool tensors."


Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)

xuhdev committed:
Update on "Let logical_xor support non-bool tensors."


Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)

xuhdev committed:
Update on "Let logical_xor support non-bool tensors."


Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)

xuhdev commented:
I noticed that this commit makes BinaryOpsKernel.cu significantly slower to compile. What could be the reason?

xuhdev committed:
Update on "Let logical_xor support non-bool tensors."


Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)

gchanan commented on test/test_torch.py:
```diff
@@ -1885,24 +1885,27 @@ def test_logical_not(self):
-        for dtype in (torch.bool,):  # Will add more dtypes in the future
+        for dtype in torch.testing.get_all_dtypes():
+            if dtype == torch.bfloat16:
+                continue
```
nit: it's generally better to do an assertRaises here so that if we add support for bfloat16, we will be forced to update the test to ensure it's tested.

Otherwise we'll just never test this.

xuhdev replied:
@gchanan  Updated!

gchanan commented on test/test_torch.py:
```diff
@@ -1885,24 +1885,27 @@ def test_logical_not(self):
-            self.assertEqual(expected_res, a)
+            for other_dtype in torch.testing.get_all_dtypes():
+                if dtype == torch.bfloat16:
+                    continue
```
here too.

gchanan commented on test/test_torch.py:
```diff
@@ -1885,24 +1885,27 @@ def test_logical_not(self):
+                # out
+                for out_dtype in torch.testing.get_all_dtypes():
+                    if dtype == torch.bfloat16:
+                        continue
```
here too.

xuhdev committed:
Update on "Let logical_xor support non-bool tensors."


Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)

xuhdev commented:
@gchanan Would you mind landing this when you feel is ready? This would significantly reduce chances of conflict while I'm continuing stacking `logical_and` and `logical_or`.

gchanan commented:
I've been trying to land it, you'll notice the tests in the first PR in the stack aren't done yet.

xuhdev commented:
@gchanan No problem! Take your time

xuhdev committed:
Update on "Let logical_xor support non-bool tensors."


Differential Revision: [D16719299](https://our.internmc.facebook.com/intern/diff/D16719299)
